### ISSUE TITLE
Cannot get autoformatting (format on save) working - VS Code

### ISSUE THREAD

**Issue by evanhfox** [hashicorp/vscode-terraform#542]
Hi There,

After ensuring my VSCode terraform extension was up to date (running 2.5.0 released an hour ago) and adding the proper settings to my settings.json vscode config file:
```
{
    "git.autofetch": true,
    "git.confirmSync": false,
    "editor.renderControlCharacters": true,
    "terraform.languageServer": {
        "[terraform]": {
            "editor.formatOnSave": true
        },
        "terraform-ls.experimentalFeatures": {
        },
        "external": true,
        "pathToBinary": "",
        "args": [
            "serve"
        ],
        "maxNumberOfProblems": 100,
        "trace.server": "off"
    },
}
```

I'm not able to get formatting to occur on a save. Is there something I'm missing? I've also added the terraform format on save settings block outside of the terraform.languageserver block, but get the same result (nothing happening to intentionally poorly formatted code with no indentions or spaces).

**Comment by nantiferov** [comment]
Yep, it stopped working a couple of weeks ago.

**Comment by tommyb82** [comment]
It stopped working for me a few days ago - v.2.5.0.  Tried reverting to previous release - no difference 😢

**Comment by aeschright** [comment]
Hi folks! Sorry you're running into this. We recently made some adjustments to how failed communication between the extension and language server are handled, so it would be helpful if you could confirm whether you're still having this issue while using the most recent extension and language server releases.

**Comment by davedash** [comment]
Hi @aeschright .

I opened a `.tf` file and removed some spaces in front of an aligned equal sign, and hit save.  Here's the output I received:

```
2021/04/14 10:28:05 server.go:581: Received 1 new requests
2021/04/14 10:28:05 server.go:174: Processing 1 requests
2021/04/14 10:28:05 server.go:265: Checking request for "workspace/executeCommand": {"command":"7.terraform-ls.rootmodules","arguments":["uri=file:///Users/davedash/Projects/terraform-config/modules/tome-backend-service/lb.tf"]}
2021/04/14 10:28:05 rpc_logger.go:29: Incoming request for "workspace/executeCommand" (ID 13): {"command":"7.terraform-ls.rootmodules","arguments":["uri=file:///Users/davedash/Projects/terraform-config/modules/tome-backend-service/lb.tf"]}
2021/04/14 10:28:05 rpc_logger.go:50: Response to "workspace/executeCommand" (ID 13): {"responseVersion":0,"doneLoading":true,"rootModules":[{"uri":"file:///Users/davedash/Projects/terraform-config/modules/tome-backend-service","name":"modules/tome-backend-service"}]}
2021/04/14 10:28:05 server.go:246: Completed 1 requests [443.057µs elapsed]
2021/04/14 10:28:06 server.go:581: Received 1 new requests
2021/04/14 10:28:06 server.go:174: Processing 1 requests
2021/04/14 10:28:06 server.go:265: Checking request for "textDocument/codeLens": {"textDocument":{"uri":"file:///Users/davedash/Projects/terraform-config/modules/tome-backend-service/lb.tf"}}
2021/04/14 10:28:06 server.go:297: Task error: [-32601] no such method "textDocument/codeLens"
2021/04/14 10:28:06 rpc_logger.go:45: Error for "textDocument/codeLens" (ID 14): [-32601] no such method "textDocument/codeLens"
2021/04/14 10:28:06 server.go:246: Completed 1 requests [265.017µs elapsed]
[Error - 10:28:06 AM] Request textDocument/codeLens failed.
  Message: no such method "textDocument/codeLens"
  Code: -32601 
```

Hope this helps.

**Comment by acdha** [comment]
I have formatting working but am seeing the same issues with the “no such method” error which is presumably what leads to the no schema popup messages:

```
[Error - 8:06:14 PM] Request textDocument/codeLens failed.
  Message: no such method "textDocument/codeLens"
  Code: -32601 
```

```
Terraform v0.15.0
on darwin_amd64
+ provider registry.terraform.io/hashicorp/aws v3.37.0
+ provider registry.terraform.io/hashicorp/dns v3.1.0
+ provider registry.terraform.io/hashicorp/http v2.1.0
```

`terraform-ls`:
```
0.15.0
platform: darwin/amd64
go: go1.16
compiler: gc
```

**Comment by ziggythehamster** [comment]
I also reproduce this. v2.1.1 _does_ work for me though.

**Comment by radeksimko** [comment]
Hi folks,
I have a hunch that one of the reasons for the functionality to be "flipping" is due to VS Code's handling of conflicts when multiple extensions claim formatting support for the same file.

In other words: If you have any other extensions supporting Terraform files - have you tried disabling those?

We plan to investigate these conflicts also as part of https://github.com/hashicorp/vscode-terraform/issues/692

You can also explicitly tell VS Code to use this extension for formatting Terraform files using the following snippet of config:

```json
    "[terraform]": {
        "editor.defaultFormatter": "hashicorp.terraform",
        "editor.formatOnSave": true
    }
```

--- 

With all that in mind, can anyone reproduce the issue on the latest version (`2.13.1`) with the latest LS version (`0.19.0`) and with other Terraform extensions (if any are installed) disabled? If so I'd be interested in setting up a debugging session with you or gathering more details.

I will leave this open for a few more weeks and then close unless I receive any confirmation of the above.

Thanks!

**Comment by acdha** [comment]
I never had another Terraform-releated extension installed but I haven't seen this problem recur in weeks.

**Comment by lukesammy** [comment]
@radeksimko I have had this issue with no other related extension installed, but upgrading to 2.13.1 with latest LS (0.19.0) it is now working! 👍 thank you

**Comment by zesty** [comment]
tl;dr  disable, re-enable Prettier -> success  #692 

Seemed not to be working for me in 2.13.2 (with the settings options set as below; default formatter set to None/null). Would either complain about no formatter for the filetype, or would do nothing. I disabled prettier in the workspace. Working after that. Re-enabling prettier hasn't ruined anything (yet).

```
    "terraform.languageServer": {
        "external": true,
        "pathToBinary": "",
        "args": [
            "serve"
        ],
        "maxNumberOfProblems": 100,
        "trace.server": "off"
    },
    "[terraform]": {
        "editor.defaultFormatter": "hashicorp.terraform",
        "editor.formatOnSave": true
    },
    "terraform-ls.experimentalFeatures": {
      "validateOnSave": true
    }
```

**Comment by radeksimko** [comment]
For anyone who managed to reproduce this problem with the latest `2.14.0+` extension I would also ask for the following details:

 - full config
 - confirmation that you managed to reproduce this problem with other (Terraform) extensions disabled
 - version of language server installed & running (this can be seen in the first few lines of the relevant `Output` `terraform-ls` pane as shown below)

![Screenshot 2021-07-28 at 18 59 22](https://user-images.githubusercontent.com/287584/127373312-6b49ec15-e460-4ac6-aed1-97245eaaf9b3.png)
![Screenshot 2021-07-28 at 18 59 07](https://user-images.githubusercontent.com/287584/127373308-97c0e7ec-e2ca-4cdf-8498-0fbb0bf0b872.png)

**Comment by audiolomb** [comment]
It stopped working for me too.

Terraform extension version: 2.14.0
Language server version is: 0.19.1

No other Terraform extensions are installed.

settings.json:

```
    "terraform.languageServer": {
        "external": true,
        "args": [
            "serve"
        ],
        "maxNumberOfProblems": 100,
        "trace.server": "off"
    },
    "[terraform]": {
        "editor.defaultFormatter": "hashicorp.terraform",
        "editor.formatOnSave": true
    },
```


Thank you,
Claudio

**Comment by radeksimko** [comment]
Thank you for the confirmation @audiolomb 

Would you be willing to jump on a brief Zoom call and debug this together? I'd really like to get to the bottom of this issue.

You can pick time here: https://calendly.com/d/nnyc-hfpp/terraform-vscode-extension-language-server-feedback

**Comment by echototo** [comment]
Just in case @audiolomb, I don't think I had the same issue but after updating the Terraform binary, my environment variable PATH was not up to date... 
@radeksimko, thank you for mentioning the output console!

**Comment by ToniCipriani** [comment]
```
2021/08/12 17:05:25 server.go:591: Received 1 new requests
2021/08/12 17:05:25 server.go:175: Processing 1 requests
2021/08/12 17:05:25 server.go:266: Checking request for "textDocument/formatting": {"textDocument":{"uri":"file:///Users/(redacted)/Documents/code/(redacted)terraform-modules/azure/azure_ad/variables.tf"},"options":{"tabSize":2,"insertSpaces":true}}
2021/08/12 17:05:25 rpc_logger.go:29: Incoming request for "textDocument/formatting" (ID 746): {"textDocument":{"uri":"file:///Users/(redacted)/Documents/code/(redacted)terraform-modules/azure/azure_ad/variables.tf"},"options":{"tabSize":2,"insertSpaces":true}}
2021/08/12 17:05:25 rpc_logger.go:45: Error for "textDocument/formatting" (ID 746): [-32098] Terraform (CLI) is required. Please install Terraform or make it available in $PATH
2021/08/12 17:05:25 server.go:247: Completed 1 requests [91.82µs elapsed]
[Error - 5:05:25 PM] Request textDocument/formatting failed.
  Message: Terraform (CLI) is required. Please install Terraform or make it available in $PATH
  Code: -32098 
```

I'm getting the same error saying that Terraform is not found, however Terraform is already installed AND accessible via PATH. FWIW Mine is installed via Brew if that actually matters, so from the path it's a symbolic link to the Brew Cellar:

```
➜  ~ which terraform
/usr/local/bin/terraform
➜  ~ ls -ltr /usr/local/bin/terraform 
lrwxr-xr-x  1 calcheu  admin  40  2 Jun 12:55 /usr/local/bin/terraform -> ../Cellar/terraform/0.15.3/bin/terraform
➜  ~ terraform --version
Terraform v0.15.3
on darwin_amd64

Your version of Terraform is out of date! The latest version
is 1.0.4. You can update by downloading from https://www.terraform.io/downloads.html
➜  ~ 
```

**Comment by radeksimko** [comment]
@ToniCipriani thank you for the report - that is useful.

I didn't realize that in this edge case we don't surface the error anywhere in the UI - we should.

Relatedly I expect we could do more in helping folks troubleshoot this by printing the whole `$PATH` somewhere somehow on demand.

Given that you're on macOS I suspect that this may be related to how VS Code gets launched.
An application which gets launched from Spotlight or the Dock doesn't necessarily inherit the same ENV variables as one which is launched from the Terminal. Can you try launching VS Code via `code /path/to/tf/dir` from the same Terminal where you confirmed `$PATH` is set correctly?

**Comment by ToniCipriani** [comment]
Believe it or not, I think this actually fixed it.

- I did the `code /usr/local/bin` as you stated
- VS Code opens, and prompts me to trust the workspace/directory
- I went ahead and trusted it, tested a tf file save with that window in another directory, and it did the format
- Restarted VSCode completely, and now other windows are formatting on save as well.

**Comment by dferramosi** [comment]
confirming it is now working for me as well! I just found this issue, and then realized i hadn't restarted vscode in a while, updating worked!

**Comment by rba1-source** [comment]
terraform needs to be in your $PATH environment variable according to the error message. For those of us who don't have access to edit this (for example, on a restricted corporate machine), how do we specify this? In v1 we could specify the path to the terraform binary in the "args" array.

**Comment by kemko** [comment]
In my case autoformatting stops working if `"editor.formatOnSaveMode": "modifications"` presents in settings.json and working if I commenting this setting.

**Comment by dferramosi** [comment]
> terraform needs to be in your $PATH environment variable

I don't see that? unless it's the %231 error? it is in my path though, and `"editor.formatOnSaveMode": "file"` works, while `"editor.formatOnSaveMode": "modifications"` does no formatting at all

**Comment by staranto** [comment]
> In my case autoformatting stops working (for both modified and unmodified lines) if `"editor.formatOnSaveMode": "modifications"` presents in settings.json and working if I commenting this setting.

@kemko hit it for me.  

I had ```editorFormatOnSaveMode: "modificationsIfAvailable"``` and it would not work.  Commented that out and I'm golden.

Version: 1.60.2
Commit: 7f6ab5485bbc008386c4386d08766667e155244e
Date: 2021-09-22T12:01:43.795Z
Electron: 13.1.8
Chrome: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Linux x64 5.11.0-34-generic snap

Launching language server: /home/xxxxxx/.vscode/extensions/hashicorp.terraform-2.15.0/lsp/terraform-ls serve
2021/09/29 06:04:27 serve_command.go:162: Starting terraform-ls 0.22.0

**Comment by dferramosi** [comment]
Well I've changed nothing in VS Code and it stopped formatting again. uggh. Only system change is upgrade to macOS 11.6

**Comment by gtback** [comment]
👍🏻 to what @kemko said:

> In my case autoformatting stops working (for both modified and unmodified lines) if "editor.formatOnSaveMode": "modifications" presents in settings.json and working if I commenting this setting.

I have `"editor.formatOnSaveMode": "modifications",` set globally in my VS Code `settings.json` file, but adding this for Terraform fixes it:

```jsonc
  "[terraform]": {
    "editor.formatOnSaveMode": "file",
  }
```

(I work in a repo with a lot of code that doesn't comply with the default settings for the formatters I use, but fortunately we run `terraform fmt` as a part of our CI.)

**Comment by sorinstanila** [comment]
Hi all, 
I was able to fix it in my case.
I found the an error in the logs, that you need to have the terraform CLI installed. Setting this to point to terraform binary fixed the formatting
```
"terraform-ls.terraformExecPath":"your path here for /bin/darwin/terraform""
```
I also added the option to format for the `file` not only `modifications`
```
"[terraform]": {
        "editor.formatOnSaveMode": "file",
  }
```

**Comment by nfantone** [comment]
Same as everyone else. `"editor.formatOnSaveMode": "modifications"` doesn't seem to be supported by the `"hashicorp.terraform"` formatter. `"editor.formatOnSaveMode": "file"` seems to works fine, though.

**Comment by jpogran** [comment]
Hello all, thank you for your continued feedback. I am writing a shorter update than I wanted to due to the end of year holidays, but did want to keep you all in the loop and summarize some of what we are seeing.

Most of you are using the settings as documented in our [readme](https://github.com/hashicorp/vscode-terraform#formatting). For those not following those settings, please review that section and try those. We have to specify this extension as the `editor.defaultFormatter` because there are other extensions that advertise the ability to format Terraform files and VS Code chooses which one depending on activation order (something outside of our control) and which is non-deterministic in general and complicated in specific cases. 

For those trying to use `editor.formatOnSaveMode`, you will have to use `file` for now. The `file` setting works because `terrafrom fmt` was written to format an entire file, not ranges. The `modifications` or `modificationsIfAvailable` settings [use the currently configured SCM](https://code.visualstudio.com/updates/v1_49#_only-format-modified-text) to detect file line ranges that have changed and send those ranges to the formatter and expect only those ranges returned. In addition, if you don't have a SCM enabled for the files you are editing, `modifications` won't work at all. The `modificationsIfAvailable` setting will fall back to `file` if there is no SCM and will appear to work sometimes.

In summary, if you want to automatically format your entire Terraform file on save, us the following configuration:

```
"[terraform]": {
  "editor.defaultFormatter": "hashicorp.terraform",
  "editor.formatOnSave": true,
  "editor.formatOnSaveMode": "file"
},
```

I did not see it mentioned here, but if you are using `editor.codeActionsOnSave` and versions of `terraform-ls` older than 0.24.0 there is a known issue. If you had `editor.codeActionsOnSave` configured with `source.fixAll` there was a [bug in our implementation](https://github.com/hashicorp/terraform-ls/issues/679) that was fixed in the [0.24.0](https://github.com/hashicorp/terraform-ls/pull/680) release. If you pin your `terraform-ls` version with `terrafrom.langaugeServer.requiredVersion`, please update to the latest as there are fixes you are missing. 

If you want to use `editor.codeActionsOnSave` with `source.formatAll.terraform` and `editor.FormatOnSave`, then I have an explanation and a recommendation for you. VS Code runs `editor.codeActionsOnSave` and `editor.FormatOnSave` separately and will not determine if a `codeAction` is performing the same operation as a formatter.  This means you can format a file more than once on save if both are configured at the same time. This is part of VS Code, not just our extension, so we don't have much control here.

So, if you want to use `editor.codeActionsOnSave` with `source.formatAll.terraform` and `editor.FormatOnSave` please use the following configuration. This will keep the global `editor.formatOnSave` for other languages you use, and configure the terraform extension to only format during a `codeAction` sweep.

```
"editor.formatOnSave": true,
"[terraform]": {
  "editor.defaultFormatter": "hashicorp.terraform",
  "editor.formatOnSave": false,
  "editor.codeActionsOnSave": {
    "source.formatAll.terraform": true
  },
},
```

I plan to write some more detailed documentation in an official place after the holidays. Hope this helps in the interim.

**Comment by staranto** [comment]
> The file setting works because terrafrom fmt was written to format an entire file, not ranges. 

Now that I see that in writing, it makes perfect sense and is obvious.  Fantastic update!  Thanks!
